### PR TITLE
Added functionality to download invoice as pdf

### PR DESCRIPTION
**Pull request by hariads**

### CODE REVIEW

File: server/puppeteerserver.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+const puppeteer = require('puppeteer');
+const express = require('express');
+const cors = require('cors');
+const fs = require('fs');
+const app = express();
+
+app.use(cors());
+
+app.get('/getPDF', async (req, res) => {
+  createHTMLFile(req.query.invoicePreview);
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: Send data in body

File: server/puppeteerserver.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  const resPDF = await generatePDF();
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: naming not clear

File: server/puppeteerserver.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const cors = require('cors');
+const fs = require('fs');
+const app = express();
+
+app.use(cors());
+
+app.get('/getPDF', async (req, res) => {
+  createHTMLFile(req.query.invoicePreview);
+  const resPDF = await generatePDF();  
+  DeleteFile(`${__dirname}/invoice.html`);
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: function names should not start with caps

File: server/puppeteerserver.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+const puppeteer = require('puppeteer');
+const express = require('express');
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: only routes should be in this file. file name should be `routes.js`

File: server/puppeteerserver.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+{
+  fs.unlink(filepath, (err) => {
+    if (err) throw err;
+  });
+}
+
+app.listen(4000, () => { console.log('listening on port 4000') });
+
+//string constants
+const htmlPrefix = '<html style="-webkit-print-color-adjust: exact;"><head><link href="https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css" rel="stylesheet"></head><body>';
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: This is wrong

File: server/puppeteerserver.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+{ 
+  const browser = await puppeteer.launch({
+    args: ['--no-sandbox', '--disable-setuid-sandbox']
+  });
+  const page = await browser.newPage();
+  await page.goto(`file:${__dirname}/invoice.html`, { waitUntil: 'networkidle0' });
+  const pdf = await page.pdf({ format: 'A4' });
+  await browser.close();
+  return pdf;
+}
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: This should be in pupeteer.js

File: server/puppeteerserver.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  res.send(resPDF);
+});
+
+function createHTMLFile(invoicePreview)
+{
+  const htmlDoc = `${htmlPrefix} ${invoicePreview} </body></html>`;    
+  fs.appendFile('invoice.html', htmlDoc, function (err) {
+    if (err) throw err;
+  }); 
+}
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: this should be in utils

File: src/components/Main.jsx [hunk 8]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
     }
   }
 
+  printPDF = () => {
+    //for temp loading
+    document.body.classList.add("cursor-wait")
```
Comment (karthikraja34) [inline src/components/Main.jsx]: This is not necessary. Use tailwind spin class `https://tailwindcss.com/docs/animation#spin`

File: src/components/Main.jsx [hunk 9]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+
+    const invoicePreview = document.getElementById('invoice').innerHTML;
```
Comment (karthikraja34) [inline src/components/Main.jsx]: Don't access DOM directly

File: src/components/Main.jsx [hunk 10]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+    return axios.get(`http://localhost:4000/getPDF`, {
```
Comment (karthikraja34) [inline src/components/Main.jsx]: URL should not be hardcoded here. Write it in util file or take it from env

Comment (karthikraja34) [reply src/components/Main.jsx]: https://create-react-app.dev/docs/adding-custom-environment-variables/

File: src/components/Main.jsx [hunk 11]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+      responseType: 'arraybuffer',
+      headers: {
+        'Accept': 'application/pdf'
+      },
+      params: { invoicePreview }
```
Comment (karthikraja34) [inline src/components/Main.jsx]:
```suggestion
      params: { invoiceHTML }
```

Comment (karthikraja34) [reply src/components/Main.jsx]: Send data in body

File: src/components/Main.jsx [hunk 12]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+    }).then((response) => {
+      document.body.classList.remove("cursor-wait")
```
Comment (karthikraja34) [inline src/components/Main.jsx]: .

File: src/components/Main.jsx [hunk 13]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+      const blob = new Blob([response.data], { type: 'application/pdf' })
+      const link = document.createElement('a')
```
Comment (karthikraja34) [inline src/components/Main.jsx:58]: Check if this can be done in reactive way

File: src/components/Main.jsx [hunk 14]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+      link.href = window.URL.createObjectURL(blob)
+      link.download = `invoice.pdf`
+      link.click()
+    })
+      .catch(err => { console.log(err) })
```
Comment (karthikraja34) [inline src/components/Main.jsx]: Handle exception

File: server/puppeteerserver.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  DeleteFile(`${__dirname}/invoice.html`);
+  res.set({ 'Content-Type': 'application/pdf', 'Content-Length': resPDF.length });
+  res.send(resPDF);
+});
+
+function createHTMLFile(invoicePreview)
+{
+  const htmlDoc = `${htmlPrefix} ${invoicePreview} </body></html>`;    
+  fs.appendFile('invoice.html', htmlDoc, function (err) {
+    if (err) throw err;
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: What will happen when error occurs

File: src/components/Main.jsx [hunk 16]
```diff
@@ -39,27 +41,53 @@ class MainComponent extends Component {
...
+                domNode.attribs.onChange = this.onChange;
+              }
+              if (domNode.name === "button") {
+                domNode.attribs.onClick = this.onClick;
+              }
+              return domNode
             }
-            return domNode
-          }
-        })}
```
Comment (karthikraja34) [inline src/components/Main.jsx]: Existing format change should not come in this PR

**Review by karthikraja34 — CHANGES_REQUESTED** [review]

File: .env [hunk 17]
```diff
@@ -0,0 +1 @@
+REACT_APP_EKMO_SERVER_ENDPOINT = http://localhost:4000
```
Comment (karthikraja34) [inline .env:1]: Env should not be committed

File: server/puppeteerserver.js (hunk 1 above)
Comment (hariads) [reply server/puppeteerserver.js]: Converted as post request

File: server/puppeteerserver.js (hunk 2 above)
Comment (hariads) [reply server/puppeteerserver.js]: gave proper namings

File: server/puppeteerserver.js (hunk 3 above)
Comment (hariads) [reply server/puppeteerserver.js]: change function name as deleteFile

File: server/puppeteerserver.js (hunk 4 above)
Comment (hariads) [reply server/puppeteerserver.js]: moved functions to utill.

File: server/puppeteerserver.js (hunk 5 above)
Comment (hariads) [reply server/puppeteerserver.js]: Changed style referance to template.

File: src/components/Main.jsx (hunk 14 above)
Comment (hariads) [reply src/components/Main.jsx]: handled

File: server/puppeteerserver.js (hunk 15 above)
Comment (hariads) [reply server/puppeteerserver.js]: Server will send 500 status code.

File: server/index.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    });
+    res.send(resultPDF);
+  }
+  catch (e) {
+    console.log(e)
+    res.status(500).send('Something broke!')
+  }
+});
+
+app.listen(4000, () => { console.log('listening on port 4000') });
```
Comment (karthikraja34) [inline server/index.js]: Port number should not be hardcoded

File: server/index.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    const filepath = fileUtil.createFile(req.body.htmlComponent, 'invoice.html');
+    const resultPDF = await generatePDF(filepath);
+    fileUtil.deleteFile(filepath);
+    res.set({
+      'Content-Type': 'application/pdf', 'Content-Length': resultPDF.length,
+      'Content-Disposition': 'inline'
+    });
+    res.send(resultPDF);
+  }
+  catch (e) {
```
Comment (karthikraja34) [inline server/index.js]: Broad Exception clause

File: server/puppeteerserver.js [hunk 20]
```diff
@@ -0,0 +1,47 @@
+const puppeteer = require('puppeteer');
+const express = require('express');
+const cors = require('cors');
+const fs = require('fs');
+const app = express();
+
+app.use(cors());
+
```
Comment (karthikraja34) [inline server/puppeteerserver.js]: This file should be removed

File: server/utils/file_utility.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
+const fs = require('fs')
```
Comment (karthikraja34) [inline server/utils/file_utility.js]: can be renamed as `file_utils`

File: src/components/Main.jsx [hunk 22]
```diff
@@ -1,16 +1,20 @@
...
 import data from "../templates/default/data.json";
+import loadspinner from "../assets/loading.svg";
 
 class MainComponent extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      "previewMode": false,
+      "previewMode": false,      
+      loading: false,
```
Comment (karthikraja34) [inline src/components/Main.jsx]: Context is explicit

File: src/components/Main.jsx [hunk 23]
```diff
@@ -39,27 +43,60 @@ class MainComponent extends Component {
...
-        </button>
+            this.setState({ previewMode: !this.state.previewMode })
+          }}>
+            Toggle Preview Mode
+          </button>
+          {
+            this.state.loading ?
+            <img src={loadspinner} className="animate-spin w-10 mx-3 inline" />
+              :                            
+            <button className="bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4 mx-3 rounded" onClick={this.printPDF}>Print</button>
```
Comment (karthikraja34) [inline src/components/Main.jsx]: Bring loading inside button

**Review by karthikraja34 — CHANGES_REQUESTED** [review]

File: server/puppeteerserver.js (hunk 6 above)
Comment (hariads) [reply server/puppeteerserver.js]: File removed

File: server/puppeteerserver.js (hunk 7 above)
Comment (hariads) [reply server/puppeteerserver.js]: File removed

File: .env (hunk 17 above)
Comment (hariads) [reply .env:1]: added in gitignore

File: server/index.js (hunk 18 above)
Comment (hariads) [reply server/index.js]: added in the env file

File: server/puppeteerserver.js (hunk 20 above)
Comment (hariads) [reply server/puppeteerserver.js]: file removed

File: src/components/Main.jsx (hunk 22 above)
Comment (hariads) [reply src/components/Main.jsx]: modified variable name

File: server/utils/file_utility.js (hunk 21 above)
Comment (hariads) [reply server/utils/file_utility.js]: renamed

File: src/components/Main.jsx (hunk 23 above)
Comment (hariads) [reply src/components/Main.jsx]: added loading inside button

**Review by karthikraja34 — APPROVED** [review]

**Pull request merged**